### DISCUSSION TITLE
Error:  About WinHttp. WinHttpRequest. 5.1

### DISCUSSION THREAD

**Discussion by junnylc** [wqweto/VbAsyncSocket#34] [Q&A] [answered]
We have written an https server using the VbAsyncSocket control and VB6 code, and accessed it using the COM "WinHttp. WinHttpRequest. 5.1". The code is as follows:

1    set xPost = CreateObject("WinHttp.WinHttpRequest.5.1")
2    xPost.Open "GET", "https://xxx.xxx.xxx/a.png", False
3    xPost.Send
4    xpost.WaitForResponse
5    WriteBinFile "x:\xxx\xx.xxx", xPost.responseBody
6    set xPost = Nothing

So there was an error on line 3.

If the second line is changed from False to True, an error occurs in the fourth line.
Using the COM "WinHttp. WinHttpRequest. 5.1" to access https services on other websites will not cause errors.

May I ask what is the reason for this?

**Comment by junnylc** [comment]
The https server has used the latest version of the VbAsyncSocket.

**Reply by wqweto** [reply]
Try to reproduce in a minimal code you can share here.

Just a small VB6 project (single form) with your TLS server which is not working w/ WinHttpRequest requests so I can debug it.

**Reply by junnylc** [reply]
> Try to reproduce in a minimal code you can share here.
> 
> Just a small VB6 project (single form) with your TLS server which is not working w/ WinHttpRequest requests so I can debug it.

```
' Start server        
   '.pvSocket.PkiPemImportRootCaCertStore appPath + "Certificate\sdlcsz.cn_root.pem"
    If HttpON Then
        Load ListenSock(1)
        With ListenSock(1)
        .Close_
        .Protocol = sckTCPProtocol
        .Bind WebPort, HttpIP
        .Listen
        End With
    End If
    
    If SSLON Then
        Load ListenSock(2)
        With ListenSock(2)
        .Close_
        .Protocol = sckTLSProtocol
        .Bind SSLPort, SSLIP
        .Listen PfxFile, ExtraPass
        '.Listen PemFile
        End With
    End If
'ListenSock_ConnectionRequest
    Load sckPost(ConNum)
    With sckPost(ConNum)
    .Accept requestID
    .Protocol = ListenSock(Index).Protocol 'SSL
    End With
```

'''''NOTE!!!,When we accessed https server using the COM "WinHttp. WinHttpRequest. 5.1" once, The ConnectionRequest event will occur twice, And the requestID of the two times is still different.


THANKS!!!

**Reply by wqweto** [reply]
Can you ZIP a working project? A simple Project1.vbp with Form1.frm with server code which fails connection from WinHttpRequest.

You don't have any http parsing in the code above. The connection is probably established and the client (WinHttpRequest) sends a request but then timeouts waiting for proper http response. Note that you have to match "http" vs "https" protocols in the `Open` call with plain TCP vs TLS enabled listener in the server.

Btw, there is a full working http server w/ TLS support in [Form2.frm](https://github.com/wqweto/VbAsyncSocket/blob/master/test/Winsock/Form2.frm) in repo's `test/Winsock` directory.

**Reply by junnylc** [reply]
how to submit ZIP document?

**Reply by wqweto** [reply]
Btw, if you do not setup a valid https certificate on your server it automatically generates (and uses) a self-signed certificate which cannot be automatically approved by WinHttpRequest (and any other certificate validating client).

To ignore certificate validation errors in WinHttpRequest you have to use its `Option` property before calling `Open` like this
```
xPost.Option(WinHttpRequestOption_SslErrorIgnoreFlags) = SslErrorFlag_Ignore_All
```
or if you don't have reference to `Microsoft WinHTTP Services` typelib in your project you can use these numeric values directly
```
xPost.Option(4) = 13056
```
before calling `xPost.Open`.

**Reply by junnylc** [reply]
I have setup a valid https certificate on my server, and I have used "xPost.Option(4) = 13056" in my project, but the error still occurred.
When I use COM "Microsoft.XMLHTTP" to access my server, errors will not occur.

**Reply by junnylc** [reply]
```
    path = "F:\360MoveData\Users\lenovo\Desktop"
    iRemote = "https://sdlcsz.cn/Images/xh.png"
    'iRemote = "https://www.baidu.com/img/flexible/logo/pc/result.png"
    iLocal = path & "\cc.png"
    set xPost = CreateObject("WinHttp.WinHttpRequest.5.1")
    xPost.Option(4) = 13056
    'set xPost = CreateObject("Microsoft.XMLHTTP")
    xPost.Open "GET", iRemote, False
    xPost.setRequestHeader("If-Modified-Since", "0")
    xPost.Send
    'xpost.WaitForResponse
    a = xPost.responseBody()
    set xPost = Nothing
```
The above code will still cause errors at "xPost.Send", If I switch to 'Microsoft.XMLHTTP', there won't be any mistakes

**Reply by wqweto** [reply]
You might want to setup WinHttpRequest to use TLS 1.2 connecting to your server like this
```
xPost.Option(WinHttpRequestOption_SecureProtocols) = SecureProtocol_ALL Or SecureProtocol_TLS1_1 Or SecureProtocol_TLS1_2
```
or in numeric 
```
xPost.Option(9) = 2728
```
before calling `Send`

This might actually be a problem with `ctxWinsock` servers supporting only TLS 1.3 and TLS 1.2 although compiled with native TLS backend which supports older versions of the protocol. I'll research this more for next release.

**Comment by junnylc** [comment]
Please try downloading [https://pan.baidu.com/s/1_-pxUld0JQsvSfuT8ohL4w](https://pan.baidu.com/s/1_-pxUld0JQsvSfuT8ohL4w)

**Reply by junnylc** [reply]
or show me your email address

**Reply by wqweto** [reply]
Cannot download this link -- requires login.

My email is all over the headers of all sources files in this repo -- you can look it up yourself.

**Reply by junnylc** [reply]
Sorry.
I have sent a ZIP document to your email

**Reply by wqweto** [reply]
Nothing received here. Try `xPost.Option(9) = 2728` as mentioned above.

Which OS version you are testing this on?

**Reply by junnylc** [reply]
The email that I sent to vlado@vis.bg was returned by the system. I can't think of a good way

**Reply by wqweto** [reply]
You can use my gmail address. It's in the header of every source file in this repository as I already mentioned.

**Reply by junnylc** [reply]
> Nothing received here. Try `xPost.Option(9) = 2728` as mentioned above.
> 
> Which OS version you are testing this on?

A Script executor, like vbs

**Comment by wqweto** [comment, marked as answer]
Btw, you can attach ZIP files to this thread by drag&drop from Explorer (e.g. [aaa.zip](https://github.com/wqweto/VbAsyncSocket/files/11573567/aaa.zip))

**Reply by junnylc** [reply]
[HttpsTLS.zip](https://github.com/wqweto/VbAsyncSocket/files/11573966/HttpsTLS.zip)
Wow, what artifact operation is this?

**Reply by wqweto** [reply]
I just tested your server with this
```
Option Explicit

Private Sub Form_Load()
    Dim xPost As Object
    Set xPost = CreateObject("WinHttp.WinHttpRequest.5.1")
    xPost.Open "GET", "https://localhost:449/a.png", False
    xPost.Option(4) = 13056
    xPost.Send
    Debug.Print UBound(xPost.ResponseBody)
End Sub
```
. . . and it seems to connect and work ok here.

Note that your `.Bind 449, "0.0.0.0"` will require setting up Windows firewall to allow remote connections.

Binding to `127.0.0.1` (localhost interface) does **not** require punching through the firewall but will allow connections from local machine only.

**Reply by junnylc** [reply]
> I just tested your server with this
> 
> ```
> Option Explicit
> 
> Private Sub Form_Load()
>     Dim xPost As Object
>     Set xPost = CreateObject("WinHttp.WinHttpRequest.5.1")
>     xPost.Open "GET", "https://localhost:449/a.png", False
>     xPost.Option(4) = 13056
>     xPost.Send
>     Debug.Print UBound(xPost.ResponseBody)
> End Sub
> ```
> 
> . . . and it seems to connect and work ok here.
> 
> Note that your `.Bind 449, "0.0.0.0"` will require setting up Windows firewall to allow remote connections.
> 
> Binding to `127.0.0.1` (localhost interface) does **not** require punching through the firewall but will allow connections from local machine only.

![1](https://github.com/wqweto/VbAsyncSocket/assets/106741778/452f3775-0e0d-40f0-b42f-8c8077123dc2)
![2](https://github.com/wqweto/VbAsyncSocket/assets/106741778/6e892e6a-688b-4bde-aeff-646cb5206eb9)

**Reply by wqweto** [reply]
I don't have your client code. With my client code above I get this
```
Https Server start
127.0.0.1 Contect - 1
Send to 127.0.0.1 - 1
127.0.0.1 Close - 1
```

**Reply by junnylc** [reply]
> I don't have your client code. With my client code above I get this
> 
> ```
> Https Server start
> 127.0.0.1 Contect - 1
> Send to 127.0.0.1 - 1
> 127.0.0.1 Close - 1
> ```

Sincerely thank you for your help! It may be an issue with my COM version, I am trying to upgrade the version

**Reply by junnylc** [reply]
> I don't have your client code. With my client code above I get this
> 
> ```
> Https Server start
> 127.0.0.1 Contect - 1
> Send to 127.0.0.1 - 1
> 127.0.0.1 Close - 1
> ```

```
Dim iRemote, xPost, a, stm, iLocal, path

path = "F:\360MoveData\Users\lenovo\Desktop"
iRemote = "https://sdlcsz.cn/Images/xh.png"
'iRemote = "https://www.baidu.com/img/flexible/logo/pc/result.png"
iLocal = path & "\cc.jpg"
set xPost = CreateObject("WinHttp.WinHttpRequest.5.1")
'set xPost = CreateObject("Microsoft.XMLHTTP")
xPost.Open "GET", iRemote, False
xPost.setRequestHeader "If-Modified-Since", "0"
xPost.Send
'xpost.WaitForResponse
a = xPost.responseBody
set xPost = Nothing

BinSave iLocal, a

sub BinSave(FileName, b)
    Dim stm
    set stm = Createobject("Adodb.Stream")
    stm.type = 1
    stm.open
    stm.write b
    stm.SaveToFile FileName, 2
    stm.close
    set stm = nothing
End sub
```
Full VBS code

**Reply by wqweto** [reply]
With your Full VBS code I get `WinHttp.WinHttpRequest: The certificate authority is invalid or incorrect`. After tweaking the code to this:
```
Dim iRemote, xPost, a, stm, iLocal, path

'path = "F:\360MoveData\Users\lenovo\Desktop"
path = "D:\TEMP"
'iRemote = "https://sdlcsz.cn/Images/xh.png"
'iRemote = "https://www.baidu.com/img/flexible/logo/pc/result.png"
iRemote = "https://127.0.0.1:449/xh.png"
iLocal = path & "\cc.jpg"
set xPost = CreateObject("WinHttp.WinHttpRequest.5.1")
'set xPost = CreateObject("Microsoft.XMLHTTP")
xPost.Open "GET", iRemote, False
xPost.Option(4) = 13056
xPost.setRequestHeader "If-Modified-Since", "0"
xPost.Send
'xpost.WaitForResponse
a = xPost.responseBody
set xPost = Nothing

BinSave iLocal, a

sub BinSave(FileName, b)
    Dim stm
    set stm = Createobject("Adodb.Stream")
    stm.type = 1
    stm.open
    stm.write b
    stm.SaveToFile FileName, 2
    stm.close
    set stm = nothing
End sub
```
. . . I get on the server this:
```
127.0.0.1 Contect - 2
Send to 127.0.0.1 - 2
127.0.0.1 Close - 2
```
Try your client code accessing `HttpsTLS.zip` server running in VB IDE on `127.0.0.1:449` not on `sdlcsz.cn` as there could be a additional reverse proxy between your external `sdlcsz.cn` address and the TLS server on your machine.
